### PR TITLE
ci: Adopt semantic release

### PR DESCRIPTION
**Pull request by haifeng-li-at-salesforce**
add semantic release bot

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -13,62 +13,87 @@ The intent and goal of open sourcing this project is to increase the contributor
...
+# Release Process
+
+### Branching Strategy:
+
+-   **Main Branch**: Ongoing development and new features are submitted to the main branch.
+
+-   **Release Branch**: We'll utilize a dedicated branch for managing releases
+
+    -   **Major Release:**
+        -   Update the `.releaserc.yml` file: Locate the branches property and specify the newly created release branch name (e.g., `release-2`).
```
Comment (ben-zhang-at-salesforce) [inline CONTRIBUTING.md:82]: should we switch line 82 with line 83? 82 line say "newly created brach ...",   83 line is for created the new branch ?

**Review by ben-zhang-at-salesforce — APPROVED** [review]
looks good beside a minor wording issue. 

npm token is set up.   we should create the release-1 branch from main --> manually run 'release" work flow to have our first release :)

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+jobs:
+    lint-build-test:
+        name: Lint, Build, Test - Node ${{ matrix.node-version }} on ${{ matrix.os }}
+
+        runs-on: ${{ matrix.os }}
+        strategy:
+            matrix:
+                node-version:
+                    - 16.x
```
Comment (khawkins) [inline .github/workflows/release.yml:15]: We only support LTS versions of Node.js. We should only have 18 and 20 in here.

Comment (haifeng-li-at-salesforce) [reply .github/workflows/release.yml:15]: Thanks for pointing this out. version 16 is removed.

File: .releaserc.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+---
+branches:
+    - release-1
+
+plugins:
+    - - '@semantic-release/commit-analyzer'
```
Comment (khawkins) [inline .releaserc.yml:6]: &lt;rant&gt;
My thought coming into this analysis: "I hate YAML." I'm not sure that's not still true, but what I think is vexing me more is the configuration schema for this particular config. **Note:** This is not your fault: I do not intend to shoot the messenger. 😉

But why in the world did the folks responsible for this schema decide that each plugin item should be a list, with the first item of the list being the name of the plugin, and the second being its configuration options? It took me _forever_ to pull apart this configuration (and I can partly thank my above-mentioned hatred of YAML for that), to come to the conclusion that this was a super long way around simply making a dictionary where the plugin name is the key and the configuration is its value. Like every other like-minded configuration structure under the sun.

I.e. why did they not make this configuration structure:

```yaml
plugins: 
-
  @semantic-release/commit-analyzer: 
    preset: "conventionalcommits"
    releaseRules: 
    - type: "docs"
      scope: "README"
      release: "patch"
-
  @semantic-release/changelog: 
    changelogTitle: "# Changelog\n\nAll notable changes to this project will be documented in this file."
```

YAML gets the continued shade for things like `- -` to denote a list of lists.
&lt;/rant&gt;

Comment (khawkins) [reply .releaserc.yml:6]: I actually have one action item to take away from this rant: Can we convert this to a `.releaserc.json` file instead? While the configuration structure is still convoluted, reading it in JSON is going to make my head explode a lot less than YAML is, every time I have to come in and edit it.

Comment (haifeng-li-at-salesforce) [reply .releaserc.yml:6]: Cannot agree more. It takes me quite a time to digest this weird data structure.

File: .releaserc.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+            - type: docs
+              scope: README
+              release: patch
+
+    - '@semantic-release/release-notes-generator'
+    - - '@semantic-release/changelog'
+      - changelogTitle: "# Changelog\n\nAll notable changes to this project will be documented in this file."
+    - - '@semantic-release/npm'
+      - npmPublish: true
+        tarballDir: 'tarball'
```
Comment (haifeng-li-at-salesforce) [inline .releaserc.yml:18]: Package tarball directory, the tarball file would be included as github asset

File: .releaserc.yml (hunk 3 above)
Comment (haifeng-li-at-salesforce) [reply .releaserc.yml:6]: Good suggestion. Let me do it now.

File: package.json [hunk 5]
```diff
@@ -62,25 +62,34 @@
...
+        "@semantic-release/git": "^10.0.1",
+        "@semantic-release/npm": "^12.0.1",
         "@types/eslint": "^8.56.10",
         "@types/estree": "^1.0.5",
+        "@types/jest": "^29.5.12",
         "@typescript-eslint/eslint-plugin": "^7.8.0",
         "@typescript-eslint/parser": "^7.8.0",
         "@typescript-eslint/rule-tester": "^7.8.0",
         "@typescript-eslint/utils": "^7.8.0",
+        "conventional-changelog-conventionalcommits": "^8.0.0",
```
Comment (haifeng-li-at-salesforce) [inline package.json:80]: Cross reference Jason Schroeder's recent fix on `release-notes-generator` update. Version `^8.0.0` on `conventional-changelog-conventionalcommits' will not cause issue. So we don't need to do anything.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -13,62 +13,87 @@ The intent and goal of open sourcing this project is to increase the contributor
...
 6. **Submit** a Pull Request against the `main` branch and refer to the issue(s) you are fixing. Try not to pollute your pull request with unintended changes. Keep it simple and small.
 7. **Sign** the Salesforce CLA (you will be prompted to do so when submitting the Pull Request)
 
 > **NOTE**: Be sure to [sync your fork](https://help.github.com/articles/syncing-a-fork/) before making a pull request.
 
+# Release Process
+
+### Branching Strategy:
+
+-   **Main Branch**: Ongoing development and new features are submitted to the main branch.
```
Comment (khawkins) [inline CONTRIBUTING.md:77]:
```suggestion
-   **Main Branch**: Ongoing development and new features are submitted to the `main` branch.
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -13,62 +13,87 @@ The intent and goal of open sourcing this project is to increase the contributor
...
+
+-   **Release Branch**: We'll utilize a dedicated branch for managing releases
```
Comment (khawkins) [inline CONTRIBUTING.md:79]:
```suggestion
-   **Release Branches**: We'll utilize dedicated branches for managing releases
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -13,62 +13,87 @@ The intent and goal of open sourcing this project is to increase the contributor
...
+
+    -   **Major Release:**
+        -   Create a new branch named `release-\*` (e.g., `release-2`) from `main` to manage the release process.
```
Comment (khawkins) [inline CONTRIBUTING.md:82]:
```suggestion
        -   Create a new branch named `release-<major version>` (e.g. `release-2` for a `2.0` release) from `main` to manage the release process.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -13,62 +13,87 @@ The intent and goal of open sourcing this project is to increase the contributor
...
+        -   Update the `.releaserc.yml` file: Locate the `branches` property and specify the newly created release branch name (e.g., `release-2`).
```
Comment (khawkins) [inline CONTRIBUTING.md:83]:
```suggestion
        -   Update the `.releaserc.json` file: Locate the `branches` property and specify the newly created release branch name (e.g. `release-2`).
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -13,62 +13,87 @@ The intent and goal of open sourcing this project is to increase the contributor
...
+    -   **Minor/Patch Release:**
+        -   If a release branch for a major version already exists (e.g., `release-2`), we can use it for subsequent minor and patch releases.
```
Comment (khawkins) [inline CONTRIBUTING.md:85]:
```suggestion
        -   We will use existing major release branches (e.g. `release-2`) for subsequent minor and patch releases (e.g. `2.0.1`, `2.1.2`, etc.).
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -13,62 +13,87 @@ The intent and goal of open sourcing this project is to increase the contributor
...
+        -   Hot fixes are merged in release branch, and will be merged back to `main` once it is verified.
```
Comment (khawkins) [inline CONTRIBUTING.md:86]:
```suggestion
        -   Once changes are verified and merged into the release branch, they will be merged back to `main`.
```

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -13,62 +13,87 @@ The intent and goal of open sourcing this project is to increase the contributor
...
+
+-   **Trigger Release Manually:**:
```
Comment (khawkins) [inline CONTRIBUTING.md:88]:
```suggestion
-   **Trigger Release Manually:**
```

File: commitlint.config.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/* This is the pattern from semantic-release */
+const automaticCommitPattern = /^chore\(release\):.*\[skip ci\]/;
+
+/** @type {import('@commitlint/types').UserConfig} */
+module.exports = {
```
Comment (khawkins) [inline commitlint.config.js:12]: Nit: `commitlint` [supports](https://commitlint.js.org/reference/configuration.html) `.mjs` and `.ts` config files. Can we move this to an ESM format?

**Review by khawkins — COMMENTED** [review]
Just some minor comments. Looks good over all!

File: commitlint.config.js (hunk 13 above)
Comment (haifeng-li-at-salesforce) [reply commitlint.config.js]: All suggested changes are adopted. Thanks for detailed review!

**Pull request merged**